File: sources/tools/stride-assets-vscode/src/core/assetParser.ts
import * as path from 'path';
import { AssetEntry } from './assetIndex';
import { findAssetReferences, findSourcePaths, AssetReferenceMatch, SourcePathMatch } from './referencePattern';

// Asset type tag (line 1): !TypeName
const TYPE_TAG_REGEX = /^!(\w+)/;
// Asset Id (line 2): Id: GUID
const ID_REGEX = /^Id:\s+([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})/i;
// Source path
const SOURCE_REGEX = /^Source:\s+(.+)/;

export interface AssetHeader {
    id: string;
    typeName: string;
    sourcePath?: string;
}

export interface ParsedAsset {
    id: string;
    typeName: string;
    sourcePath?: string;
    references: AssetReferenceMatch[];
    sourcePaths: SourcePathMatch[];
    parts: AssetPartDefinition[];
}

export interface AssetPartDefinition {
    id: string;
    name?: string;
    line: number;
}

// Fast header-only parse: reads just the first few lines to get Id and type
export function parseAssetHeader(content: string): AssetHeader | null {
    const lines = content.split('\n', 10); // Only need first few lines
    if (lines.length < 2) {
        return null;
    }

    const typeMatch = TYPE_TAG_REGEX.exec(lines[0]);
    if (!typeMatch) {
        return null;
    }

    const idMatch = ID_REGEX.exec(lines[1]);
    if (!idMatch) {
        return null;
    }

    let sourcePath: string | undefined;
    for (let i = 2; i < Math.min(lines.length, 10); i++) {
        const sourceMatch = SOURCE_REGEX.exec(lines[i]);
        if (sourceMatch) {
            const val = sourceMatch[1].trim();
            if (!val.startsWith('!') && val !== 'null') {
                sourcePath = val;
            }
            break;
        }
    }

    return {
        id: idMatch[1].toLowerCase(),
        typeName: typeMatch[1],
        sourcePath,
    };
}

// Full parse: extracts all references, source paths, and entity definitions
export function parseAssetFull(content: string): ParsedAsset | null {
    const header = parseAssetHeader(content);
    if (!header) {
        return null;
    }

    const references = findAssetReferences(content);
    const sourcePaths = findSourcePaths(content);
    const parts = parseAssetParts(content);

    // Filter out the asset's own Id from references (it appears on line 1 sometimes)
    const filteredRefs = references.filter(r => r.guid !== header.id);

    return {
        ...header,
        references: filteredRefs,
        sourcePaths,
        parts,
    };
}

// Parse asset part definitions (entities in scenes/prefabs, UI elements in pages/libraries)
// Parts are defined as indented blocks with an Id: GUID field
const PART_ID_REGEX = /^(\s+)Id:\s+([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})/i;
const PART_NAME_REGEX = /^\s+Name\*?:\s+(.+)/;

function parseAssetParts(content: string): AssetPartDefinition[] {
    const parts: AssetPartDefinition[] = [];
    const lines = content.split('\n');
    // The asset's own Id is on line 1 - skip it
    // Part Ids appear deeper in the hierarchy (indented)
    for (let i = 2; i < lines.length; i++) {
        const idMatch = PART_ID_REGEX.exec(lines[i]);
        if (idMatch && idMatch[1].length > 0) { // Must be indented (not the root Id)
            const part: AssetPartDefinition = {
                id: idMatch[2].toLowerCase(),
                line: i,
            };
            // Look ahead for a Name field among sibling properties.
            // Id: and Name: are at the same indentation level (siblings).
            // Stop when we hit a line at lesser indentation (left the parent block).
            const idIndent = idMatch[1].length;
            for (let j = i + 1; j < lines.length; j++) {
                const lineJ = lines[j];
                // Skip empty lines
                if (lineJ.trim().length === 0) {
                    continue;
                }
                // Measure indentation — if below the Id line's level, we've left the block
                const lineIndent = lineJ.length - lineJ.trimStart().length;
                if (lineIndent < idIndent) {
                    break;
                }
                const nameMatch = PART_NAME_REGEX.exec(lineJ);
                if (nameMatch) {
                    part.name = nameMatch[1].trim();
                    break;
                }
            }
            parts.push(part);
        }
    }
    return parts;
}

// Build an AssetEntry from a parsed header and file path
export function headerToAssetEntry(header: AssetHeader, filePath: string): AssetEntry {
    const ext = path.extname(filePath);
    const baseName = path.basename(filePath, ext);
    return {
        id: header.id,
        location: baseName,
        filePath,
        typeName: header.typeName,
        extension: ext,
        sourcePath: header.sourcePath,
    };
}

File: sources/tools/stride-assets-vs/AssetParser.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace StrideAssets.VisualStudio
{
    internal sealed class AssetEntry
    {
        public AssetEntry(string id, string typeName, string filePath, string? sourcePath)
        {
            Id = id;
            TypeName = typeName;
            FilePath = filePath;
            SourcePath = sourcePath;
        }

        public string Id { get; }
        public string TypeName { get; }
        public string FilePath { get; }
        public string? SourcePath { get; }
    }

    internal sealed class AssetPartEntry
    {
        public AssetPartEntry(string id, string filePath, int line, string? name, string? ownerName = null)
        {
            Id = id;
            FilePath = filePath;
            Line = line;
            Name = name;
            OwnerName = ownerName;
        }

        public string Id { get; }
        public string FilePath { get; }
        public int Line { get; }
        /// <summary>Entity name, set for entity-level parts.</summary>
        public string? Name { get; }
        /// <summary>Owning entity name, set for component-level parts.</summary>
        public string? OwnerName { get; }
    }

    internal sealed class BackReference
    {
        public BackReference(string sourceFilePath, int line, int column, string context)
        {
            SourceFilePath = sourceFilePath;
            Line = line;
            Column = column;
            Context = context;
        }

        public string SourceFilePath { get; }
        public int Line { get; }
        public int Column { get; }
        public string Context { get; }
    }

    internal sealed class ParsedAsset
    {
        public ParsedAsset(string id, string typeName, string? sourcePath,
            IReadOnlyList<(string guid, string name)> references,
            IReadOnlyList<AssetPartEntry> parts)
        {
            Id = id;
            TypeName = typeName;
            SourcePath = sourcePath;
            References = references;
            Parts = parts;
        }

        public string Id { get; }
        public string TypeName { get; }
        public string? SourcePath { get; }
        public IReadOnlyList<(string guid, string name)> References { get; }
        public IReadOnlyList<AssetPartEntry> Parts { get; }
    }

    /// <summary>
    /// Regex-based parser for Stride YAML asset files (.sd*).
    /// Ported from the TypeScript assetParser.ts / referencePattern.ts in stride-assets.
    /// </summary>
    internal static class AssetParser
    {
        // Top-level asset type tag: first non-empty line, e.g. "!MaterialAsset"
        private static readonly Regex TypeTagRegex = new Regex(
            @"^!([A-Za-z_][\w.]*(?:,\s*[A-Za-z_][\w.]*)?)",
            RegexOptions.Compiled | RegexOptions.Multiline);

        // Top-level Id line, e.g. "Id: f16ed18b-ed49-4471-ab10-bb4ee742bdbb"
        private static readonly Regex IdLineRegex = new Regex(
            @"^\s*Id:\s+([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})",
            RegexOptions.Compiled | RegexOptions.Multiline);

        // Asset reference: GUID:AssetName (may contain spaces in asset name)
        private static readonly Regex AssetRefRegex = new Regex(
            @"([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}):([^\s,}\]]+(?:\s+[^\s,}\]]+)*)",
            RegexOptions.Compiled);

        // Source: path (raw resource path for textures, models, etc.)
        private static readonly Regex SourcePathRegex = new Regex(
            @"^\s*Source:\s+(.+)",
            RegexOptions.Compiled | RegexOptions.Multiline);

        private const int HeaderReadLength = 1000;

        /// <summary>
        /// Fast header parse — examines at most the first 1000 chars.
        /// Extracts type tag (line 0), Id (within first ~10 lines), optional Source.
        /// Returns null if no Id found (not a proper Stride asset).
        /// </summary>
        public static AssetEntry? ParseHeader(string content, string filePath)
        {
            var header = content.Length > HeaderReadLength
                ? content.Substring(0, HeaderReadLength)
                : content;

            var typeMatch = TypeTagRegex.Match(header);
            if (!typeMatch.Success)
                return null;

            var idMatch = IdLineRegex.Match(header);
            if (!idMatch.Success)
                return null;

            var id = idMatch.Groups[1].Value.ToLowerInvariant();
            var typeName = typeMatch.Groups[1].Value;

            string? sourcePath = null;
            var sourceMatch = SourcePathRegex.Match(header);
            if (sourceMatch.Success)
                sourcePath = sourceMatch.Groups[1].Value.Trim();

            return new AssetEntry(id, typeName, filePath, sourcePath);
        }

        /// <summary>
        /// Full file parse: extracts all asset references (excluding the asset's own Id)
        /// and all part definitions (entities/elements in composite assets).
        /// </summary>
        public static ParsedAsset? ParseFull(string content, string filePath)
        {
            var entry = ParseHeader(content, filePath);
            if (entry == null)
                return null;

            var refs = new List<(string guid, string name)>();
            foreach (Match m in AssetRefRegex.Matches(content))
            {
                var guid = m.Groups[1].Value.ToLowerInvariant();
                if (guid == entry.Id)
                    continue;
                refs.Add((guid, m.Groups[2].Value.Trim()));
            }

            var parts = ParseParts(content, filePath);

            string? sourcePath = null;
            var sourceMatch = SourcePathRegex.Match(content);
            if (sourceMatch.Success)
                sourcePath = sourceMatch.Groups[1].Value.Trim();

            return new ParsedAsset(entry.Id, entry.TypeName, sourcePath, refs, parts);
        }

        /// <summary>
        /// Extracts part definitions from composite assets (scenes, prefabs, UI pages).
        /// Parts are identified by indented Id: lines (2+ leading spaces).
        /// Looks ahead up to 5 lines for a sibling Name: field at the same indentation.
        /// </summary>
        public static List<AssetPartEntry> ParseParts(string content, string filePath)
        {
            var parts = new List<AssetPartEntry>();
            var lines = content.Split('\n');

            // Track the current entity context to assign OwnerName to component-level parts.
            // The first Id: indent seen is the entity level; deeper indents are components.
            int? entityIndent = null;
            string? currentEntityName = null;

            for (int i = 0; i < lines.Length; i++)
            {
                var line = lines[i];

                // Must be indented (part Id lines have 2+ leading spaces, not top-level)
                if (line.Length < 2 || line[0] != ' ' || line[1] != ' ')
                    continue;

                var idMatch = IdLineRegex.Match(line);
                if (!idMatch.Success)
                    continue;

                var partId = idMatch.Groups[1].Value.ToLowerInvariant();
                var indent = line.Length - line.TrimStart().Length;

                bool isEntity = (entityIndent == null || indent <= entityIndent);
                if (isEntity)
                {
                    entityIndent = indent;

                    // Look ahead for sibling Name: at the same indentation
                    string? name = null;
                    for (int j = i + 1; j < Math.Min(i + 6, lines.Length); j++)
                    {
                        var next = lines[j];
                        if (string.IsNullOrWhiteSpace(next))
                            continue;
                        var nextIndent = next.Length - next.TrimStart().Length;
                        if (nextIndent != indent)
                            break;
                        var trimmed = next.TrimStart();
                        if (trimmed.StartsWith("Name:", StringComparison.Ordinal))
                        {
                            name = trimmed.Substring(5).Trim();
                            break;
                        }
                        break; // different field at same indent — no Name sibling
                    }

                    currentEntityName = name;
                    parts.Add(new AssetPartEntry(partId, filePath, i + 1, name, ownerName: null));
                }
                else
                {
                    // Component-level Id — no Name: field exists in YAML for components.
                    parts.Add(new AssetPartEntry(partId, filePath, i + 1, name: null, ownerName: currentEntityName));
                }
            }

            return parts;
        }
    }
}
